File: src/Cordango.Standalone/Security/EntityAccess.cs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) Cordango and contributors.
// Part of Cordango, the open application language and compiler: https://github.com/cordango/cordango
// Licensed under the Apache License, Version 2.0. See LICENSE in the repository root.

namespace Cordango.Standalone.Security;

/// <summary>How one field may be touched: <c>Read</c> = it appears in responses, <c>Update</c> = it
/// may be set on create and update.</summary>
public sealed record FieldRule(bool Read, bool Update);

/// <summary>
/// One caller's effective access to one entity: the union of every role they hold, resolved down to
/// entity-level CRUD plus the fields that differ from it.
///
/// <para>Immutable and computed from data alone, so it can be built once per request and asked
/// anything without touching the database or the HTTP context.</para>
/// </summary>
public sealed class EntityAccess
{
    private readonly IReadOnlyDictionary<string, FieldRule> _fields;
    private readonly IReadOnlySet<string> _commands;
    private readonly bool _allCommands;

    public EntityAccess(
        bool create, bool read, bool update, bool delete,
        IReadOnlyDictionary<string, FieldRule>? fields = null,
        IReadOnlySet<string>? commands = null,
        bool allCommands = false)
    {
        Create = create;
        Read = read;
        Update = update;
        Delete = delete;
        _fields = fields ?? EmptyFields;
        _commands = commands ?? EmptyCommands;
        _allCommands = allCommands;
    }

    public bool Create { get; }
    public bool Read { get; }
    public bool Update { get; }
    public bool Delete { get; }

    private static readonly IReadOnlyDictionary<string, FieldRule> EmptyFields =
        new Dictionary<string, FieldRule>(StringComparer.Ordinal);

    private static readonly IReadOnlySet<string> EmptyCommands = new HashSet<string>(StringComparer.Ordinal);

    /// <summary>An administrator: everything, including every command.</summary>
    public static readonly EntityAccess Full = new(true, true, true, true, allCommands: true);

    /// <summary>Nothing at all — no role of the caller's says anything about this entity.</summary>
    public static readonly EntityAccess None = new(false, false, false, false);

    /// <summary>What a caller gets on an application whose definition declares no roles: they may
    /// look, and that is all. An application that has decided nothing about access should not have
    /// write access decided for it.</summary>
    public static readonly EntityAccess ReadOnly = new(false, true, false, false);

    public bool Allows(string operation) => operation switch
    {
        "create" => Create,
        "read" => Read,
        "update" => Update,
        "delete" => Delete,
        _ => false,
    };

    /// <summary>A field with no override inherits the entity-level answer.</summary>
    public bool CanReadField(string field) => _fields.TryGetValue(field, out var rule) ? rule.Read : Read;

    public bool CanUpdateField(string field) => _fields.TryGetValue(field, out var rule) ? rule.Update : Update;

    /// <summary>Fields to strip from responses.</summary>
    public IEnumerable<string> HiddenReadFields => _fields.Where(kv => !kv.Value.Read).Select(kv => kv.Key);

    /// <summary>Fields a write may not touch. Keyed off explicit overrides only, so a role that may
    /// create records can still set every ordinary field — the restriction is the exception, not the
    /// default.</summary>
    public IEnumerable<string> WriteRestrictedFields => _fields.Where(kv => !kv.Value.Update).Select(kv => kv.Key);

    /// <summary>Deny-by-default. True only for an administrator, or when some grant of the caller's
    /// names this command (or <c>"*"</c>).</summary>
    public bool CanRunCommand(string commandKey) => _allCommands || _commands.Contains(commandKey);

    /// <summary>The command keys allowed, for the client's own enablement decisions. <c>"*"</c> when
    /// every command is allowed.</summary>
    public IReadOnlyCollection<string> AllowedCommands => _allCommands ? ["*"] : [.. _commands];
}

File: src/Cordango.Standalone.Node/src/security/entity-access.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) Cordango and contributors.
// Part of Cordango, the open application language and compiler: https://github.com/cordango/cordango
// Licensed under the Apache License, Version 2.0. See LICENSE in the repository root.

/** How one field may be touched: `read` = it appears in responses, `update` = it may be set on
 * create and update. */
export type FieldRule = { read: boolean; update: boolean };

const emptyFields: ReadonlyMap<string, FieldRule> = new Map();
const emptyCommands: ReadonlySet<string> = new Set();

/**
 * One caller's effective access to one entity: the union of every role they hold, resolved down to
 * entity-level CRUD plus the fields that differ from it.
 *
 * Immutable and computed from data alone, so it can be built once per request and asked anything
 * without touching the database or the request context.
 */
export class EntityAccess {
  private readonly fields: ReadonlyMap<string, FieldRule>;
  private readonly commands: ReadonlySet<string>;
  private readonly allCommands: boolean;

  constructor(
    readonly create: boolean,
    readonly read: boolean,
    readonly update: boolean,
    readonly delete_: boolean,
    fields?: ReadonlyMap<string, FieldRule>,
    commands?: ReadonlySet<string>,
    allCommands = false,
  ) {
    this.fields = fields ?? emptyFields;
    this.commands = commands ?? emptyCommands;
    this.allCommands = allCommands;
  }

  /** An administrator: everything, including every command. */
  static readonly full = new EntityAccess(true, true, true, true, undefined, undefined, true);

  /** Nothing at all — no role of the caller's says anything about this entity. */
  static readonly none = new EntityAccess(false, false, false, false);

  /** What a caller gets on an application whose definition declares no roles: they may look, and
   * that is all. */
  static readonly readOnly = new EntityAccess(false, true, false, false);

  allows(operation: string): boolean {
    switch (operation) {
      case "create":
        return this.create;
      case "read":
        return this.read;
      case "update":
        return this.update;
      case "delete":
        return this.delete_;
      default:
        return false;
    }
  }

  /** A field with no override inherits the entity-level answer. */
  canReadField(field: string): boolean {
    return this.fields.get(field)?.read ?? this.read;
  }

  canUpdateField(field: string): boolean {
    return this.fields.get(field)?.update ?? this.update;
  }

  /** Fields to strip from responses. */
  get hiddenReadFields(): string[] {
    return [...this.fields.entries()].filter(([, rule]) => !rule.read).map(([field]) => field);
  }

  /** Fields a write may not touch. Keyed off explicit overrides only, so the restriction is the
   * exception, not the default. */
  get writeRestrictedFields(): string[] {
    return [...this.fields.entries()].filter(([, rule]) => !rule.update).map(([field]) => field);
  }

  /** Deny-by-default. True only for an administrator, or when some grant of the caller's names this
   * command (or `"*"`). */
  canRunCommand(commandKey: string): boolean {
    return this.allCommands || this.commands.has(commandKey);
  }

  /** The command keys allowed, for the client's own enablement decisions. `["*"]` when every
   * command is allowed. */
  get allowedCommands(): string[] {
    return this.allCommands ? ["*"] : [...this.commands].sort();
  }
}
